### PR TITLE
PoW refactoring - part I

### PR DESCRIPTION
**Pull request by curie-kief**
Steps to test:
- Create a password protected wallet on the old version
- Compile
- Verify that the client syncs (can be done by temporarily pointing it to a different data dir)
- Sync / change back data dir and verify that the client accepts new blocks after being synced
- Verify that the old wallet is accessible

----

This PR implements structures necessary for supporting two PoW sizes simultaneously without actually changing the underlying PoW

### CODE REVIEW

File: src/crypto/cn_slow_hash.hpp [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+{
+	val[0] = 0;
+	val[1] = 0;
+	val[2] = 0;
+	val[3] = 0;
+
+#ifdef _MSC_VER
+	__cpuidex(val, eax, ecx);
+#else
+	__cpuid_count(eax, ecx, val[0], val[1], val[2], val[3]);
```
Comment (quangvu3) [inline src/crypto/cn_slow_hash.hpp]: It appears not compatible with MingW

File: src/crypto/cn_slow_hash.hpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+	cn_slow_hash()
+	{
+#if !defined(WIN_MEM_ALIGN)
+		lpad.as_void = aligned_alloc(4096, MEMORY);
+		spad.as_void = aligned_alloc(4096, 4096);
+#else
+		lpad.as_void = _aligned_malloc(MEMORY, 4096);
+		spad.as_void = _aligned_malloc(4096, 4096);
+#endif
+		printf("cn_slow_hash allocated %.16lx %.16lx %.16lx!\n", uint64_t(this), uint64_t(lpad.as_void), uint64_t(spad.as_void));
```
Comment (quangvu3) [inline src/crypto/cn_slow_hash.hpp]: Less warning on MingW and other x86 platforms while compiling with:

`printf("cn_slow_hash allocated %.16llx %.16llx %.16llx!\n", uint64_t(this), uint64_t(lpad.as_void), uint64_t(spad.as_void));`

Is it for debug only?

File: src/crypto/cn_slow_hash.hpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+		return *this;
+	}
+
+	// Copying is going to be really inefficient
+	cn_slow_hash(const cn_slow_hash& other) = delete;
+	cn_slow_hash& operator= (const cn_slow_hash& other) = delete;
+
+	~cn_slow_hash()
+	{
+		printf("cn_slow_hash deallocated %.16lx %.16lx %.16lx!\n", uint64_t(this), uint64_t(lpad.as_void), uint64_t(spad.as_void));
```
Comment (quangvu3) [inline src/crypto/cn_slow_hash.hpp]: Less warning on MingW and other x86 platforms while compiling with:

`printf("cn_slow_hash deallocated %.16llx %.16llx %.16llx!\n", uint64_t(this), uint64_t(lpad.as_void), uint64_t(spad.as_void));`

Comment (sumogr) [reply src/crypto/cn_slow_hash.hpp]: -I get a perfect warning/error free build on linux (Ubuntu 16.04 LTS). Daemon synced, restored old wallet successfully

-Failed to build on MingW (Windows 2016 Server) at first. Tried with @quangvu3 fixes, i got it to build binaries although I get lots of warnings https://pastebin.com/aCPEppib.
Daemon sis syncing, got a wallet restored as well with no issues

**Comment by sumogr** [comment]
-I get a perfect warning/error free build on linux (Ubuntu 16.04 LTS). Daemon synced, restored old wallet successfully

-Failed to build on MingW (Windows 2016 Server) at first. Tried with @quangvu3 fixes, i got it to build binaries although I get lots of warnings https://pastebin.com/aCPEppib.
Daemon is syncing, got a wallet restored as well with no issues

File: src/crypto/cn_slow_hash.hpp (hunk 2 above)
Comment (curie-kief) [reply src/crypto/cn_slow_hash.hpp]: Yes. Sorry for pushing it.

**Comment by fireice-uk** [comment]
Thanks for your feedback. e2bde8e should build on Windows. I decided in the end to use the Windows APIs instead of working against them

**Comment by curie-kief** [comment]
Thanks for your feedback. e2bde8e supports Windows APIs fully. This should fix the build problems

**Comment by sumogr** [comment]
Yeap, I get working binaries now with MingW64 on Win 10.  
Good job @curie-kief

**Comment by quangvu3** [comment]
Yes, both Win32 and ARMv7 (32bit) got crashed when running daemon binaries while compiled successfully

**Comment by sumogr** [comment]
Confirm bug fix on Windows Server 2008 32 bit build with mingw32.
Binaries are working

**Comment by curie-kief** [comment]
The crash was because of differently defined `DataLength` for jh it is `ull` while skein is `size_t`. My recommendation would be to move all the auxiliary hashes to a single file and a single interface later.

**Comment by quangvu3** [comment]
Win32 and AMRv7 are working well now 👍

**Pull request merged**